compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignTiles.cpp

```diff
@@ -0,0 +1,437 @@
+          return WalkResult::interrupt();
+        }
+
+        SmallVector<AMDAIE::TileOp> tiles =
+            llvm::map_to_vector(logicalObjectFifo.getTiles(), [](Value tile) {
+              return dyn_cast_if_present<TileOp>(tile.getDefiningOp());
+            });
+        AMDAIE::TileOp assignedTileOp =
+            *std::min_element(tiles.begin(), tiles.end(), tileLocAndUsageCmp);
+
+        // Increase usage of the chosen tile.
+        int64_t col = getConstantIndexOrAssert(assignedTileOp.getCol());
+        int64_t row = getConstantIndexOrAssert(assignedTileOp.getRow());
+        tileLocToUsage[TileLoc(col, row)] += 1;
+
+        rewriter.setInsertionPoint(logicalObjectFifo);
+        SmallVector<Value> tileResults = {
+            cast<Value>(assignedTileOp.getResult())};
+        rewriter.replaceOpWithNewOp<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+            logicalObjectFifo,
+            cast<LogicalObjectFifoType>(
+                logicalObjectFifo.getOutput().getType()),
+            logicalObjectFifo.getMemref(), tileResults);
+        return WalkResult::advance();
+      });
+  if (res.wasInterrupted()) return failure();
+  return success();
+}
+
+namespace {
+
+class AMDAIEAssignTilesPass
+    : public impl::AMDAIEAssignTilesBase<AMDAIEAssignTilesPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AMDAIEDialect>();
+  }
+
+  void runOnOperation() override;
+};
+
+void AMDAIEAssignTilesPass::runOnOperation() {
+  Operation *parentOp = getOperation();
+  IRRewriter rewriter(&getContext());
+  auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(parentOp);
+  std::optional<AMDAIEDevice> maybeDevice = getConfigAMDAIEDevice(targetAttr);
+  if (!maybeDevice) {
+    parentOp->emitOpError()
+        << "has no AMDAIEDevice in the target attribute configuration. This "
+           "device-specific information is required to determine when loops "
+           "can be subsumed into DMA operations, and must be attached to a "
+           "containing ModuleOp.";
+    return signalPassFailure();
+  }
+  AMDAIEDeviceModel deviceModel = getDeviceModel(maybeDevice.value());
+
+  // Assign tile locations to logical objectfifos on local (L1) memory.
+  if (failed(assignLocalTiles(rewriter, parentOp))) {
+    parentOp->emitOpError() << "local tile assignment failed";
+    return signalPassFailure();
+  }
+  if (failed(verify(parentOp, true))) {
+    return signalPassFailure();
+  }
+  LLVM_DEBUG(llvm::dbgs() << "After assignLocalTiles: \n" << *parentOp << "\n");
+
+  // Duplicate global objFifos for each strided copy-like operation user to
+  // allow global logical objectFifos to be assigned to different tile
+  // locations.
+  if (failed(duplicateGlobalObjFifos(rewriter, parentOp))) {
+    parentOp->emitOpError() << "failed duplicating global object fifos";
+    return signalPassFailure();
+  }
+  if (failed(verify(parentOp, true))) {
```

Comment:
> Question 1: The additional prints and verifications: I know they're copied from another pass, but do you think maybe they're not needed here? Because they're not in other passes. They suggest to me that this pass is still not to be considered stable, is that accurate?

Yeah, good point, I think this part is actually stable and they can be removed.

> Question 2: I assume duplicateGlobalObjFifos is something you explicitly do not want to use in DistributeCoresAndObjectFifos` ? If that's not the case, you could have moved this copied code into another util function.

The reason for this was mainly to keep the behaviour of `DistributeCoresAndObjectFifos` as close as possible to what it was to avoid more test changes. Maybe this could be tried and changed in a follow-up.